command_line/dose_analysis.py

```diff
@@ -0,0 +1,306 @@
+            elif ma.info().labels == ["I", "SIGI"]:
+                intensities = ma
+            elif ma.info().labels == ["I(+)", "SIGI(+)", "I(-)", "SIGI(-)"]:
+                intensities = ma
+        if not intensities:
+            raise KeyError("Intensity array not found in mtz file")
+        if not batches:
+            raise KeyError("Batch array not found in mtz file")
+
+        indices = mtz_object.extract_original_index_miller_indices()
+        intensities = intensities.customized_copy(indices=indices)
+        batches = batches.customized_copy(indices=indices)
+
+        if params.anomalous:
+            intensities = intensities.as_anomalous_array()
+            batches = batches.as_anomalous_array()
+
+        if dose is None:
+            dose = batches_to_dose(batches.data(), params.dose)
+        else:
+            dose = dose.data()
+
+        return cls(intensities, dose, params)
+
+    @classmethod
+    def from_dials_datafiles(cls, params, experiments, reflection_table):
+        """Initialise the class from an experiment list and reflection table.
+
+        Args:
+            params: A dose-analysis phil params object
+            experiments: An ExperimentList
+            reflection_table: A reflection table.
+        """
+        reflection_table = filter_reflection_table(
+            reflection_table, intensity_choice=["scale"], partiality_threshold=0.4
+        )
+
+        # get scaled intensities
+        intensities = miller.array(
+            miller.set(
+                crystal.symmetry(
+                    unit_cell=median_unit_cell(experiments),
+                    space_group=experiments[0].crystal.get_space_group(),
+                ),
+                indices=reflection_table["miller_index"],
+                anomalous_flag=params.anomalous,
+            ),
+            data=reflection_table["intensity.scale.value"],
+            sigmas=reflection_table["intensity.scale.variance"] ** 0.5,
+        )
+        intensities.set_observation_type_xray_intensity()
+
+        doses = flex.double()
+        start_doses, doses_per_image = interpret_images_to_doses_options(
+            params, experiments
+        )
+        logger.info(
+            "Interpreting data using:\n  starting_doses=%s\n  dose_per_image=%s",
+            ", ".join("%s" % i for i in start_doses)
+            if len(set(start_doses)) > 1
+            else " all %s" % str(start_doses[0]),
+            ", ".join("%s" % i for i in doses_per_image)
+            if len(set(doses_per_image)) > 1
+            else " all %s" % str(doses_per_image[0]),
+        )
+
+        for expt, starting_dose, dose_per_img in zip(
+            experiments, start_doses, doses_per_image
+        ):
+            refls = reflection_table.select(expt)
+            imgno = flex.ceil(refls["xyzobs.px.value"].parts()[2])
+            dose = (imgno * dose_per_img) + starting_dose
+            doses.extend(dose)
+
+        doses = doses.iround()
```

Comment:
Alternatively if `dose_per_img` can only be integer, then `doses` can be a `flex.int` from the beginning.

Reply:
Thanks for pointing this out, I was going to allow floats so that actual (calculated) doses could be given, but then found that pychef wants integer doses, so best to constrain to ints for now.